lib/flatware/rspec/example_job_builder.rb

```diff
@@ -0,0 +1,131 @@
+  module RSpec
+    # groups examples into one job per worker.
+    # reads from persisted example statuses, if available,
+    # and attempts to ballence the jobs accordingly.
+    class ExampleJobBuilder
+      attr_reader :args, :workers
+
+      def initialize(args, workers:)
+        @args = args
+        @workers = workers
+
+        load_configuration_and_examples
+      end
+
+      def jobs
+        timed_examples, untimed_examples = timed_and_untimed_examples
+        bucket_count = [@examples_to_run.size, workers].min
+
+        filter_empty_jobs(
+          balance_jobs(
+            bucket_count: bucket_count,
+            timed_examples: timed_examples,
+            untimed_examples: untimed_examples
+          )
+        )
+      end
+
+      private
+
+      def balance_jobs(bucket_count:, timed_examples:, untimed_examples:)
+        buckets = Array.new(bucket_count) { Bucket.new }
+
+        timed_examples.sort_by!(&:last).reverse_each do |(example_id, time)|
+          buckets.min_by(&:runtime).add_example(example_id, time)
+        end
+
+        untimed_examples.each_with_index do |example_id, index|
+          buckets[index % bucket_count].add_example(example_id, 0)
+        end
+
+        buckets.map { |bucket| Job.new(bucket.examples, args) }
+      end
+
+      # do not return empty jobs, otherwise if there are fewer examples than workers,
+      # the workers that received empty jobs would run the whole suite
+      def filter_empty_jobs(jobs)
+        jobs.reject { |job| job.id.empty? }
+      end
+
+      def timed_and_untimed_examples
+        @examples_to_run.reduce([[], []]) do |(timed, untimed), example_id|
+          if (time = example_runtimes[example_id])
+            [timed + [[example_id, time]], untimed]
+          else
+            [timed, untimed + [example_id]]
+          end
+        end
+      end
+
+      def load_persisted_example_statuses
+        ::RSpec::Core::ExampleStatusPersister.load_from(
+          @example_status_persistence_file_path || ''
+        )
+      end
+
+      def example_runtimes
+        @example_runtimes ||= load_persisted_example_statuses.each_with_object({}) do |status_entry, runtimes|
+          next unless status_entry.fetch(:status) =~ /pass/i
+
+          runtimes[status_entry[:example_id]] = status_entry[:run_time].to_f
+        end
+      end
+
+      def load_configuration_and_examples
+        configuration = ::RSpec.configuration
+        configuration.define_singleton_method(:command) { 'rspec' }
+
+        ::RSpec::Core::ConfigurationOptions.new(args).configure(configuration)
+
+        @example_status_persistence_file_path = configuration.example_status_persistence_file_path
+
+        # Load spec files in a fork to avoid polluting the parent process,
+        # otherwise the actual execution will return warnings for redefining constants
+        # and shared example groups.
+        @examples_to_run = within_forked_process { load_examples_to_run(configuration) }
```

Comment:
This was a bit tricky - I need to access `RSpec.world.ordered_example_groups` which only works after `configuration.load_spec_files`, but calling that loads rails_helper (etc) which can define some constants, and then the actual execution in workers would load these files producing warnings.

Would be great if there was another way to do it, but for now the least-bad idea I had was to get the list of examples from a forked process.